src/particles/integrators/Integrators.cpp

```diff
@@ -0,0 +1,163 @@
+
+        // initialize numerical integration parameters
+        amrex::ParticleReal const dz = (zout-zin)/nsteps;
+        amrex::ParticleReal const alpha = 1.0_prt - pow(2.0_prt,1.0/3.0);
+        amrex::ParticleReal const tau2 = dz/(1.0_prt + alpha);
+        amrex::ParticleReal const tau1 = tau2/2.0_prt;
+        amrex::ParticleReal const tau3 = alpha*tau1;
+        amrex::ParticleReal const tau4 = (alpha - 1.0_prt)*tau2;
+
+        // initialize the value of the independent variable
+        amrex::ParticleReal zeval = zin;
+
+        // loop over integration steps
+        for(int j=0; j < nsteps; ++j)
+        {
+            map1(tau1,refpart,zeval,element_type);
+            map2(tau2,refpart,zeval,element_type);
+            map1(tau3,refpart,zeval,element_type);
+            map2(tau4,refpart,zeval,element_type);
+            map1(tau3,refpart,zeval,element_type);
+            map2(tau2,refpart,zeval,element_type);
+            map1(tau1,refpart,zeval,element_type);
+        }
+
+    }
+
+    void map1 (amrex::ParticleReal const tau, RefPart & refpart,
+                           amrex::ParticleReal & zeval, std::string element_type)
+    {
+
+        using namespace amrex::literals; // for _rt and _prt
+
+        if(element_type == "rfcavity"){
+
+           amrex::ParticleReal const t = refpart.t;
+           amrex::ParticleReal const pt = refpart.pt;
+
+           if(pt < -1.0_prt){
+              refpart.t = t + tau/sqrt(1.0_prt - pow(pt,-2));
+              refpart.pt = pt;
+           }
+           else {
+             refpart.t = t;
+             refpart.pt = pt;
+           }
+           zeval = zeval + tau;
+        }
+
+    }
+
+    void map2 (amrex::ParticleReal const tau, RefPart & refpart,
+                           amrex::ParticleReal & zeval, std::string element_type)
+    {
+
+        using namespace amrex::literals; // for _rt and _prt
+
+        if(element_type == "rfcavity"){
+
+           amrex::ParticleReal const t = refpart.t;
+           amrex::ParticleReal const pt = refpart.pt;
+           amrex::ParticleReal const k = 27.245985285371864_prt;
+           amrex::ParticleReal const phi = -0.2_prt;
+           amrex::ParticleReal const E0 = 62.0_prt;
+
+           refpart.t = t;
+           refpart.pt = pt - tau*E0*RF_Efield(zeval)*cos(k*t+phi);
+           zeval = zeval;
+           amrex::AllPrintToFile("rfout") << zeval << " " << RF_Efield(zeval) << "\n";
+        }
+    }
+
+    amrex::ParticleReal
+    RF_Efield (amrex::ParticleReal & zeval)
+    {
+        using namespace amrex::literals; // for _rt and _prt
+
+        // specify constants
+        amrex::ParticleReal const pi = 3.141592653589793_prt;
+        amrex::ParticleReal const zlen = 1.31879807_prt;
+        amrex::ParticleReal const zmid = zlen/2.0_prt;
+        int const ncoef = 25;
+
+        // specify Fourier coefficients
+        amrex::ParticleReal cos_coef [ncoef] = {};
+        amrex::ParticleReal sin_coef [ncoef] = {};
+
```

Comment:
Below is a model for the 1.3 GHz 9-cell TESLA RF cavity:

https://journals.aps.org/prab/pdf/10.1103/PhysRevSTAB.3.092001

Reply:
@cemitch99 should we add that reference in the example as a comment or so?